spring-boot/admin/src/main/java/eu/coatrack/admin/service/OauthUserAccountManagement.java

```diff
@@ -0,0 +1,119 @@
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.List;
+
+@Service
+public class OauthUserAccountManagement {
+
+    private static final String GITHUB_API_EMAIL = "https://api.github.com/user/emails";
+
+    @Autowired
+    private OAuth2AuthorizedClientService clientService;
+
+    private OAuth2User getLoggedInUser() {
+        return (OAuth2User) SecurityContextHolder
+                .getContext()
+                .getAuthentication()
+                .getPrincipal();
+    }
+
+    public String getLoginNameFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("login");
+    }
+
+    public String getNameFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("name");
+    }
+
+    public String getCompanyFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("company");
+    }
+
+    public String getEmailFromLoggedInUser() throws JsonProcessingException {
+        String email = getLoggedInUser().getAttribute("email");
+
+        if (wasGitHubMailAddressSetToPrivateAndThereforeNotDirectlyReadable(email)) {
+            ResponseEntity<String> emailListFromGithub = getEmailsListFromGithub();
+            email = getPrimaryEmailFromLoggedInUser(emailListFromGithub);
+        }
+        return email;
+    }
+
+    private boolean wasGitHubMailAddressSetToPrivateAndThereforeNotDirectlyReadable(String email) {
+        return (email == null || email.isEmpty());
+    }
+
+    private ResponseEntity<String> getEmailsListFromGithub() {
+        RestTemplate restTemplate = new RestTemplate();
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Authorization", "token " + getTokenFromLoggedInUser());
+        HttpEntity<String> githubRequest = new HttpEntity(headers);
+
+        return restTemplate.exchange(GITHUB_API_EMAIL, HttpMethod.GET,
+                githubRequest, String.class);
+    }
+
+    private String getPrimaryEmailFromLoggedInUser(ResponseEntity<String> userEmailsRequest) throws JsonProcessingException {
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+        CollectionType mapResponseWithEmailsListToAListOfGithubEmails = objectMapper.getTypeFactory().constructCollectionType(List.class, GithubEmail.class);
```

Comment:
Is it maybe possible to put the lines 96 to 99 into the Constructor or a PostConstruct method, so that they are executed only once?

Reply:
I read the whole code again, and found a way to remove the whole ObjectMapper, and get directly a List<GithubEmail> with the rest template, so there is no more need to make any PostConstruct